### PR TITLE
FlexibleDecay input block

### PR DESCRIPTION
**Pull request by wkotlarski**
I've added the `FlexibleDecay` block. For now it looks like this
```
  Block FlexibleDecay    
     0   1                    # calculate decays (0 = no, 1 = yes)    
     1   1                    # include higher order corrections in decays (0 = no, 1 = yes)
```
The `FlexibleDecay_settings` class shares a lot of similarities with `Spectrum_generator_settings` class. Should we create an extra layer of abstraction?

Bonus point questions:
1. Why do we use `enum` and not `enum class` for settings?
2. Why do we have our own version of `to_string` instead of using `std::to_string`?

### CODE REVIEW

**Comment by Expander** [comment]
I like the code you've written!

You're right, it is very similar to `Spectrum_generator_settings`. An abstraction would be nice, if it possible and does not complicate the code a lot.

Concerning your questions:

1. I've used a C-style enum initially as conversion to the underlying type (`int`) was easier (which we need to do as the enum fields correspond to numbers in the SLHA input). If you think an `enum class` works the same here, I'm not against changing it.

2. `std::to_string` produces some incovenient output format for floating point digits, sometimes it even chops off some digits, i.e. it does not allow one to print a floating point number with full precision. For this reason I've added `flexiblesusy::to_string`, which produces a better output for doubles due to `boost::lexical_cast`. Also, in some cases this is faster. :)

**Comment by pathron** [comment]
I don't mind a mind further abstraction as we might add more such classes later, but its also not necessary just now, its not much code and extra abstraction has its own downsides.  

I guess we will sort out what exact flags we offer once the structure is fixed?

**Comment by wkotlarski** [comment]
Flags can be discussed already now. The potential flags we might want to take under consideration are

- higher order SM QED
- higher order SM QCD
- higher order SM EW
- higher order BSM (wont do anything now)
- only on-shell, single off-shell, double off-shell decays of h->VV (double off-shell won't do anything now)
- no loop-induced decays

Let me know which one you'd like to have and in which order. Florian has for the new module the following flags
```
Block DECAYOPTIONS   # Options to turn on/off specific decays     
1    1     # Calc 3-Body decays of Fu     
2    1     # Calc 3-Body decays of Fe     
3    1     # Calc 3-Body decays of Fd     
1001 1     # Loop Decay of Fu     
1002 1     # Loop Decay of Fe     
1003 1     # Loop Decay of Fd     
1004 1     # Loop Decay of hh     
1114 1.     # U-factors (0: off, 1:p2_i=m2_i, 2:p2=0, p3:p2_i=m2_1)     
1115 1.     # Use loop-corrected masses for external states     
1116 0.     # OS values for W,Z and fermions (0: off, 1:g1,g2,v 2:g1,g2,v,Y_i)     
1117 0.     # Use defined counter-terms     
1118 1.     # Use everywhere loop-corrected masses for loop-induced decays    
```

**Comment by pathron** [comment]
I like the flags you suggest and I think the order you put is already the right one.  I was going to suggest "no loop-induced decays" goes first, but actually I think its less useful / obvious than switchng off higher order corrections and if its not first it should be last.

**Comment by wkotlarski** [comment]
I was not sure about the 'no loop-induced decays' flag. It's more of a debug flag so that If those decays fail, you could still get the non-loop-induced ones.

**Comment by pathron** [comment]
Yeah so I assumed there could be cases where someone does not trust loop induced decays or thinks they should be ignored, so may want to switch them off.

**Comment by Expander** [comment]
I think we should not follow SARAH's flag settings, as they seem confusing to me.

> The potential flags we might want to take under consideration are
> 
>     * higher order SM QED
>     * higher order SM QCD
>     * higher order SM EW
>     * higher order BSM (wont do anything now)
>     * only on-shell, single off-shell, double off-shell decays of h->VV (double off-shell won't do anything now)
>     * no loop-induced decays
>     * ...

This looks like a reasonable first idea, but I have some thoughts/questions:

* What exactly is the difference between SM QED and SM EW? Intuitively I'd say that EW already contains QED. I'd suggest to divide the loop corrections into: SM QED, SM Weak, SM QCD. This seems to be a distinction that can be defined more precisely.

* If we split the SM corrections into QED, Weak, QCD, then we may do the same for BSM. Or is there a reason to not do this distinction there? Is it maybe easier to generate all BSM corrections at once?

* I would agree with Peter: The "no loop-induced decays" should be first, if we introduce it.

Sorry for too much thinking, but I somehow still have the feeling these flags could be more systematic. Intuitively I think one could define the flags according to the following criteria:

1. number of loops
1. orders of couplings
1. leading/ next-to leading/ next-to-next-to leading/ ...

# number of loops

If we would use the number of loops for the flags, it seems we'd need two flags
* loop order of SM corrections (0, 1, 2, 3, ...)
* loop order of BSM corrections (0, 1, 2, 3, ...)

# orders of couplings

If we would use the couplings as a distinction criteria, we could have:
* order of SM QED corrections (0, 1, 2, 3, ...)
* order of SM Weak corrections (0, 1, 2, 3, ...)
* order of SM QCD corrections (0, 1, 2, 3, ...)
* order of BSM QED corrections (0, 1, 2, 3, ...)
* order of BSM Weak corrections (0, 1, 2, 3, ...)
* order of BSM QCD corrections (0, 1, 2, 3, ...)
* order of BSM Rest corrections (0, 1, 2, 3, ...)
Here the "order" (0, 1, 2, ...) could stand for the powers of coupling factors or so. But as these couplings mix at some high order, this seems like a bad idea.

# leading/ next-to leading/ ...

If we would use the "leadingness" a distinction criteria, we could have something like:
* order SM (0, 1, 2, ...)
* order BSM (0, 1, 2, ...)
where "order = 0" is the lowest order were a non-zero contribution exists.

The first suggestion of Wojtex seems to be a mix of these criteria. For example: What does "higher order QED" stand for? For somebody who does not know the details it could mean:
* By default FS calculates everything at tree-level. So, higher-order means 1-loop.
* If I activate 1-loop decays in FS, then "higher-order" means 2-loop.

Sorry for my unclear thoughts. I just want to have the clearest possible design and at the same time allow for the later inclusion higher loop/coupling orders w/o introducing new flags. :)

**Comment by pathron** [comment]
Alex I'd be fine with the structure you propose, though I am not sure splitting up the BSM corrections by order makes as much sense for BSM corrections that we obtain via auto generation as it does for SM ones added by hand.  So I'd also be fine with just one flag for BSM as Wojtex proposed.

Is 'weak' really any more clear than electroweak.  I feel like that would only be the case if the SM gauge group was SU(3)_C \times SU(2)_W \times U(1)_Q, which it isn't.  QED stil comntains the T_3 from SU(2).  Maybe what we mean is pure QED, EW and QCD .  I don't care much though I think people will figure out what we mean regardless.

Leading / next-to-leading fits with what we actually do for the BSM, but may be a bit counterintuaitive.  probably fine if the comment in the slha writes:
 "# 0 = L0, 1 = NLO, 2=NNLO.. so if there is no tree-level process 0 will lead to the leading order loop one-loop correction being used" and we say similar everywhere that is relevant in the manual.

**Comment by wkotlarski** [comment]
Time to finish this. `FlexibleDecay` block now works. We have
```
Block FlexibleDecay    
   0   1                    # calculate decays (0 = no, 1 = yes)    
   1   1                    # include higher order corrections in decays (0 = no, 1 = yes)    
   2   2                    # off-shell decays into VV pair    
   3   1e-5               # min br. to print
```
The remaining thing is possible revamping of flag 2. Do we have a consensus how to proceed? Are there some other flags you'd like to see?

**Review by Expander — COMMENTED** [review]
I would start a discussion on the flag choice concerning extensibility. :)
~~~
Block FlexibleDecay    
   0   1                    # calculate decays (0 = no, 1 = yes)    
   1   1                    # include higher order corrections in decays (0 = no, 1 = yes)    
   2   2                    # off-shell decays into VV pair    
   3   1e-5                 # min br. to print
~~~

* Flag 0 switching on/off the decays seems fine to me.
* Flag 1 to switch on/off the ho corrections could be generalized: 0 = LO, 1 = NLO, 2 = NNLO, 3 = NNNLO. This would allow the study of individual ho corrections in a bit more detail.
* Flag 2: Why is this flag needed? If we have it, should we not also have a flag for decays into scalar pairs and fermion pairs etc?
* Flag 3: seems fine to me, but we should probably move it to another position 

I'm wondering how this flag block would look like when we extend the decays to also include fermions and vectors in the initial state.

**Comment by wkotlarski** [comment]
Regarding 3: we did implement only offshell decays to VV because of their phenomenological importance.

**Comment by Expander** [comment]
That's of course fine. But that may change over time. ;)

File: src/FlexibleDecay_settings.cpp [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with FlexibleSUSY.  If not, see
+// <http://www.gnu.org/licenses/>.
+// ====================================================================
+
+#include "FlexibleDecay_settings.hpp"
```
[the remark this answers is not in the document]
Comment (wkotlarski) [reply src/FlexibleDecay_settings.cpp]: Done.

**Comment by pathron** [comment]
If flag 2 is set 1 we get one off-shell and one on-shell and if it’s zero we get both on-shell?  Or it really only works when it’s set to 2 currently?  I am fine either way just wanted to check.

**Comment by wkotlarski** [comment]
It's more complicated
- 0: no offshell decays
- 1: double offshell for mH < mV, single offshell for mV < mH < 2mV
- 2: double offshell for mH < 2mV

**Comment by wkotlarski** [comment]
Now we have this
```
  Block FlexibleDecay    
     0   1                    # calculate decays (0 = no, 1 = yes)    
     1   1e-5               # min br. to print    
     2   1                    # include higher order corrections in decays (0 = LO, 1 = NLO, 2 = NNLO, 3 = NNNLO)    
     3   1                    # use Thomson alpha instead of alpha(m) in decays to gamgam and gamZ    
     4   2                    # off-shell decays into VV pair  
```
Are you ok with merging into the decay branch?

**Comment by Expander** [comment]
Very nice, thanks!

However, I'd propose to slightly adapt the descriptions:
~~~
Block FlexibleDecay    
   0   1                    # calculate decays (0 = no, 1 = yes)    
   1   1e-5                 # minimum BR to print    
   2   1                    # include higher order corrections in decays (0 = LO, 1 = NLO, 2 = NNLO, 3 = NNNLO)    
   3   1                    # use Thomson alpha(0) instead of alpha(m) in decays to γγ and γZ    
   4   2                    # off-shell decays into VV pair
~~~

**Comment by wkotlarski** [comment]
How about now?

**Comment by Expander** [comment]
Perfect! :)

**Review by pathron — APPROVED** [review]
Looks good to me, thanks!

**Comment by wkotlarski** [comment]
@Expander, for you too?

**Comment by Expander** [comment]
Sorry, I can't see it: Did you adapt the file `templates/LesHouches.in` accordingly to reflect the default settings?

**Comment by wkotlarski** [comment]
Should the FlexibleDecay block be added to `template/LesHouches.in` before or after `@formattedSLHAInputBlocks@`?

**Comment by Expander** [comment]
I'd say before.

**Comment by Expander** [comment]
Could you please indent the `FlexibleDecay` block by 1 space? (So it aligns with the blocks before.) :)

**Review by Expander — APPROVED** [review]

**Comment by Expander** [comment]
Thanks a lot!

**Pull request merged**